# **Old Car Price Prediction**

---




#**Objective**
Old car price prediction is a task of estimating the market value of a used car based on various features, such as the brand, model, year, fuel, transmission,etc.

So, the main objective to create the model that can predict the price of old cars.

#**Data Source**
The dataset used in this project was downloaded from [github]('https://github.com/YBIFoundation/Dataset/raw/main/Car%20Price.csv')

Target variable(y):


*   Selling_Price

Eight features(X):


1. Brand
2. Model
3. Year
4. KM Driven
5. Fuel
6. Seller Type
7. Transmission
8. Owner







#**Import Library**

In [ ]:
import pandas as pd
import numpy as np

#**Import Data**

In [ ]:
car =  pd.read_csv('https://github.com/YBIFoundation/Dataset/raw/main/Car%20Price.csv')

In [ ]:
car.head()

Brand                     Model  Year  Selling_Price  KM_Driven    Fuel  \
0   Maruti             Maruti 800 AC  2007          60000      70000  Petrol   
1   Maruti  Maruti Wagon R LXI Minor  2007         135000      50000  Petrol   
2  Hyundai      Hyundai Verna 1.6 SX  2012         600000     100000  Diesel   
3   Datsun    Datsun RediGO T Option  2017         250000      46000  Petrol   
4    Honda     Honda Amaze VX i-DTEC  2014         450000     141000  Diesel   

  Seller_Type Transmission         Owner  
0  Individual       Manual   First Owner  
1  Individual       Manual   First Owner  
2  Individual       Manual   First Owner  
3  Individual       Manual   First Owner  
4  Individual       Manual  Second Owner

In [ ]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4340 entries, 0 to 4339
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Brand          4340 non-null   object
 1   Model          4340 non-null   object
 2   Year           4340 non-null   int64 
 3   Selling_Price  4340 non-null   int64 
 4   KM_Driven      4340 non-null   int64 
 5   Fuel           4340 non-null   object
 6   Seller_Type    4340 non-null   object
 7   Transmission   4340 non-null   object
 8   Owner          4340 non-null   object
dtypes: int64(3), object(6)
memory usage: 305.3+ KB


#**Describe Data**

In [ ]:
car.describe()

Year  Selling_Price      KM_Driven
count  4340.000000   4.340000e+03    4340.000000
mean   2013.090783   5.041273e+05   66215.777419
std       4.215344   5.785487e+05   46644.102194
min    1992.000000   2.000000e+04       1.000000
25%    2011.000000   2.087498e+05   35000.000000
50%    2014.000000   3.500000e+05   60000.000000
75%    2016.000000   6.000000e+05   90000.000000
max    2020.000000   8.900000e+06  806599.000000

In [ ]:
car.columns

Index(['Brand', 'Model', 'Year', 'Selling_Price', 'KM_Driven', 'Fuel',
       'Seller_Type', 'Transmission', 'Owner'],
      dtype='object')

#**Data Preprocessing**

In [ ]:
car['Brand'].nunique()

29

In [ ]:
car['Brand'].value_counts()

Maruti           1280
Hyundai           821
Mahindra          365
Tata              361
Honda             252
Ford              238
Toyota            206
Chevrolet         188
Renault           146
Volkswagen        107
Skoda              68
Nissan             64
Audi               60
BMW                39
Fiat               37
Datsun             37
Mercedes-Benz      35
Jaguar              6
Mitsubishi          6
Land                5
Volvo               4
Ambassador          4
Jeep                3
MG                  2
OpelCorsa           2
Daewoo              1
Force               1
Isuzu               1
Kia                 1
Name: Brand, dtype: int64

In [ ]:
car['Brand'] == 'Maruti'

0        True
1        True
2       False
3       False
4       False
        ...  
4335    False
4336    False
4337     True
4338    False
4339    False
Name: Brand, Length: 4340, dtype: bool

In [ ]:
car['Brand'] == 'Hyundai'

0       False
1       False
2        True
3       False
4       False
        ...  
4335     True
4336     True
4337    False
4338     True
4339    False
Name: Brand, Length: 4340, dtype: bool

In [ ]:
df = car.loc[(car['Brand'] == 'Maruti')|(car['Brand'] == 'Hyundai')]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2101 entries, 0 to 4338
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Brand          2101 non-null   object
 1   Model          2101 non-null   object
 2   Year           2101 non-null   int64 
 3   Selling_Price  2101 non-null   int64 
 4   KM_Driven      2101 non-null   int64 
 5   Fuel           2101 non-null   object
 6   Seller_Type    2101 non-null   object
 7   Transmission   2101 non-null   object
 8   Owner          2101 non-null   object
dtypes: int64(3), object(6)
memory usage: 164.1+ KB


In [ ]:
df['Brand'].value_counts()

Maruti     1280
Hyundai     821
Name: Brand, dtype: int64

In [ ]:
df['Fuel'].value_counts()

Petrol    1414
Diesel     627
CNG         38
LPG         22
Name: Fuel, dtype: int64

In [ ]:
df['Transmission'].value_counts()

Manual       1972
Automatic     129
Name: Transmission, dtype: int64

In [ ]:
df['Seller_Type'].value_counts()

Individual          1624
Dealer               401
Trustmark Dealer      76
Name: Seller_Type, dtype: int64

In [ ]:
df['Owner'].value_counts()

First Owner             1370
Second Owner             538
Third Owner              143
Fourth & Above Owner      50
Name: Owner, dtype: int64

#**Define Target Variable (y) and Feature Variables (X)**

In [ ]:
df.columns

Index(['Brand', 'Model', 'Year', 'Selling_Price', 'KM_Driven', 'Fuel',
       'Seller_Type', 'Transmission', 'Owner'],
      dtype='object')

In [ ]:
y=df['Selling_Price']
X = df[['Brand','Year', 'KM_Driven', 'Fuel',
       'Seller_Type', 'Transmission', 'Owner']]

#**Data Encoding**

In [ ]:
X = pd.get_dummies(X)

#**Train Test Split**

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =  train_test_split(X,y, random_state = 2529)

In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((1575, 17), (526, 17), (1575,), (526,))

#**Modeling**

In [ ]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor()

In [ ]:
model.fit(X_train,y_train)

RandomForestRegressor()

#**Prediction**

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
y_pred

array([ 760000.        ,  198904.16666667,  903593.33333333,
        473949.95      ,   82595.        ,  225000.        ,
        218838.44316667,  199100.        ,  233729.49      ,
        110285.05555556,  576334.16666667,  136678.57142857,
        726661.16666667,  349069.99      ,  507351.19047619,
        530605.83333333,  139779.98      ,  169716.66666667,
        352252.62626263,  141921.62698413,  170323.65666667,
        288700.99021717,   88292.77761905,   62885.        ,
         94630.        ,  531296.66666667,  551938.33333333,
        335753.33333333,  109173.07142857,  445336.28571429,
        737703.33333333,  540500.83333333,  178564.26571429,
        310973.9       ,  341594.14371032,  311232.54616667,
        225000.        ,  105176.30952381,  235248.7       ,
        283552.66805952,   82595.        ,  232124.96      ,
        297632.4       ,  718580.        ,  154375.        ,
        467850.        ,  337342.55799756,   93620.        ,
        418547.5       ,

#**Model Evaluation**

In [ ]:
from sklearn.metrics import mean_absolute_percentage_error,r2_score

In [ ]:
mean_absolute_percentage_error(y_test,y_pred)

0.2726632500658131

In [ ]:
r2_score(y_test,y_pred)

0.6755187879030141

#**Explanation**

By performing Random Forest Regressor model, we aim to get a better result or less error with max accuracy. Our purpose was to predict the price of the old cars having 4340 data entries.

Initially, we read the data from dataset which is available on github (Ybi Foundation).

Now we perform data preprocessing, Firstly we remove the data from brand (Skoda, Volvo, Jeep, etc) with all information because the given samples are not sufficient to perform the task.

In our dataset, 6 features are categorical variables and 2 numerical variables (price column excluded). To apply the ML model, we need to transform these categorical variables into numerical variables.

Next step is to define the target variable(y) and feature variable(X) and then we encode the data into numerical values.

Next step is to split the dataset into training and testing sets. The training set will be used to train the machine learning model, while the testing set will be used to evaluate its performance.

Now by using the Random Forest Regressor model we fit the dataset in the model which is called Data Modeling.

And then we evaluate the trained model on the testing set using appropriate metrics like mean absolute percentage error, r2 score to measure its performance.